### PR TITLE
Update BearerTokenAuthenticationPolicy implementations with sync workflow

### PR DESCRIPTION
**Pull request by samvaity**
This PR adds sync workflow to policies implementing the `BearerTokenAuthenticationPolicy`.

### CODE REVIEW

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryCredentialsPolicy.java [hunk 1]
```diff
@@ -153,6 +155,89 @@ public Mono<Boolean> authorizeRequestOnChallenge(HttpPipelineCallContext context
...
+        if (httpResponse.getStatusCode() == 401 && authHeader != null) {
+            if (authorizeRequestOnChallengeSync(context, httpResponse)) {
+                return nextPolicy.processSync();
+                // Both Netty and OkHttp expect the requestBody to be closed after the connection is closed.
+                // Failure to do so results in memory leak.
+                // In case of StreamResponse (or other scenarios where we do not eagerly read the response)
+                // we let the client close the connection after the stream read.
+                // This can cause potential leaks in the scenarios like above, where the policy
+                // may intercept the response and prevent it from reaching the client.
+                // Hence, the policy needs to ensure that the connection is closed.
```
Comment (kasobol-msft) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryCredentialsPolicy.java:206]: we should call to httpResponse.close() here.

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryTokenService.java [hunk 2]
```diff
@@ -80,7 +84,12 @@ public Mono<AccessToken> getToken(TokenRequestContext tokenRequestContext) {
...
 
             return this.refreshTokenCache.getToken(requestContext)
-                .flatMap(refreshToken -> this.tokenService.getAcrAccessTokenAsync(refreshToken.getToken(), scope, serviceName, TokenGrantType.REFRESH_TOKEN));
+                .flatMap(refreshToken -> this.tokenService.getAcrAccessTokenAsync(refreshToken.getToken(), scope,
+                    serviceName, TokenGrantType.REFRESH_TOKEN));
         }).doOnError(err -> logger.error("Could not fetch the ACR error token.", err));
     }
+
+    public AccessToken getTokenSync(TokenRequestContext tokenRequestContext) {
+        return this.getToken(tokenRequestContext).block();
```
Comment (kasobol-msft) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryTokenService.java:93]: we should have todo/work item to revisit this later. this one seems to be simpler than identity.

File: sdk/containerregistry/azure-containers-containerregistry/src/test/java/com/azure/containers/containerregistry/ContainerRegistryCredentialPolicySyncTests.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class ContainerRegistryCredentialPolicySyncTests {
```
Comment (kasobol-msft) [inline sdk/containerregistry/azure-containers-containerregistry/src/test/java/com/azure/containers/containerregistry/ContainerRegistryCredentialPolicySyncTests.java:33]: should this use @SyncAsyncTest and test both ways?

**Review by kasobol-msft — COMMENTED** [review]
lgtm

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryTokenService.java (hunk 2 above)
Comment (samvaity) [reply sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryTokenService.java:93]: https://github.com/Azure/azure-sdk-for-java/issues/29746

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryCredentialsPolicy.java [hunk 4]
```diff
@@ -153,6 +155,90 @@ public Mono<Boolean> authorizeRequestOnChallenge(HttpPipelineCallContext context
...
+     */
+    @Override
+    public void setAuthorizationHeaderSync(HttpPipelineCallContext context, TokenRequestContext tokenRequestContext) {
+        AccessToken token = tokenService.getTokenSync(tokenRequestContext);
+        context.getHttpRequest().getHeaders().set(AUTHORIZATION, BEARER + " " + token.getToken());
+    }
+
+    @Override
+    public HttpResponse processSync(HttpPipelineCallContext context, HttpPipelineNextSyncPolicy next) {
+        if ("http".equals(context.getHttpRequest().getUrl().getProtocol())) {
```
Comment (lmolkova) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryCredentialsPolicy.java:183]: is it possible to extract common code from sync and async methods?

File: sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/implementation/KeyVaultCredentialPolicy.java [hunk 5]
```diff
@@ -224,15 +322,10 @@ public String getTenantId() {
         }
     }
 
-    public static void clearCache() {
-        CHALLENGE_CACHE.clear();
-    }
-
```
Comment (vcolin7) [inline sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/implementation/KeyVaultCredentialPolicy.java:230]: I don't think we should get rid of this. Maybe it was removed on accident? I can see it on the other 3 Key Vault policy files.

**Review by vcolin7 — COMMENTED** [review]
Everything looks good to me. I just left a comment about a small section of code being removed from the KV Administration policy :)

File: sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/implementation/KeyVaultCredentialPolicy.java (hunk 5 above)
Comment (samvaity) [reply sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/implementation/KeyVaultCredentialPolicy.java:230]: I think I just moved it a couple of lines above here https://github.com/Azure/azure-sdk-for-java/pull/29730/files#diff-7bb81c7c946d6a6c3a465a01cc22b5b702607792a9f79d2c27d3efb3705d016aR287

**Review by vcolin7 — APPROVED** [review]

**Review by lmolkova — APPROVED** [review]

**Pull request merged**